File: src/codes/no-undo-history/py/rfc_memento.py
class EditorMemento:
    def __init__(self, state: str):
        self._state = state

    def get_state(self) -> str:
        return self._state

class TextEditor:
    def __init__(self):
        self._content = ""

    def type(self, text: str):
        self._content += text

    def get_content(self) -> str:
        return self._content

    def create_memento(self) -> EditorMemento:
        return EditorMemento(self._content)

    def restore(self, memento: EditorMemento):
        self._content = memento.get_state()

# 利用例
editor = TextEditor()
history: list[EditorMemento] = []

editor.type("Hello ")
history.append(editor.create_memento())

editor.type("World")
history.append(editor.create_memento())

print(editor.get_content())  # Hello World

# Undo
history.pop()  # 最新の状態を破棄
last = history[-1]
editor.restore(last)

print(editor.get_content())  # Hello 

File: src/codes/no-undo-history/ts/rfc_memento.ts
// Memento
class EditorMemento {
  constructor(private readonly state: string) {}

  getState(): string {
    return this.state;
  }
}

// Originator（状態を持つ）
class TextEditor {
  private content = "";

  type(text: string): void {
    this.content += text;
  }

  getContent(): string {
    return this.content;
  }

  createMemento(): EditorMemento {
    return new EditorMemento(this.content);
  }

  restore(memento: EditorMemento): void {
    this.content = memento.getState();
  }
}

// Caretaker（履歴管理）
const editor = new TextEditor();
const history: EditorMemento[] = [];

editor.type("Hello ");
history.push(editor.createMemento());

editor.type("World");
history.push(editor.createMemento());

console.log(editor.getContent()); // Hello World

// Undo
history.pop(); // 現在の状態は捨てる
const last = history[history.length - 1];
editor.restore(last);

console.log(editor.getContent()); // Hello
